openfermioncirq/gates/fermionic_simulation.py

```diff
@@ -86,11 +88,85 @@ def state_swap_eigen_component(x: str, y: str, sign: int = 1, angle: float=0):
     return component
 
 
+@cirq.value_equality(approximate=True)
+class FermionicSimulationGate(cirq.EigenGate):
+    """The Jordan-Wigner transform of exp(-i H) for a fermionic Hamiltonian
+    H."""
```

Comment:
On further thought, I am not sure the premise of this gate makes sense. The fermionic Hamiltonian has a representation in terms of fermionic creation and annihilation operators. But the Jordan-Wigner transform of this Hamiltonian depends on an ordering of all the fermionic modes in the system. In particular, it is impossible to specify the unitary of this gate without also specifying all modes present and their ordering. Since the JWT introduces Z-strings, the unitary may end up involving many more qubits than the original fermionic modes that the fermionic gate acts on!

Reply:
This gate is quite important and is used in many of our algorithms. But the key point you are mentioning Kevin is that it assumes a JW transform and that the gate we are implementing acts on orbitals that are adjacent in the canonical JW ordering. So in other words, it is generated by n_p n_{p +1}, a^\dagger_p a_{p+1} + cc, n_p, n_p+1 under JW. Make sense?

Reply:
The gate you're talking about is the one called FSimGate in Cirq. The gate that @bryano is proposing here is something else. Apparently the idea is to represent the JW transform of _any_ fermionic unitary, and I'm not sure that makes sense, at least not the way it's implemented right now.

Reply:
The idea is definitely not to represent the JW transform of any fermionic unitary. The goal here is specifically to refactor the three existing gates (`QuadraticFermionicSimulationGate`, `CubicFermionicSimulationGate`, `QuarticFermionicSimulationGate`), so that they have a common interface and share code to the extent possible.

Reply:
And, as @babbush noted, these gates are specifically the JW transform of exponentials of fermionic terms that act on modes that are adjacent in the JW string.

Reply:
I understand that QuadraticFermionicSimulationGate, CubicFermionicSimulationGate, QuarticFermionicSimulationGate act on adjacent modes, so the Z-strings cancel out. My question is: what class of gates is FermionicSimulationGate supposed to encompass? According to the one-line docstring you have here, it can encompass any fermionic unitary.

Reply:
It's intended to encompass those three.

Reply:
We definitely shouldn't have a public class whose sole purpose is to be a parent class to those three classes. If the only purpose is to reduce code duplication, then it should be a private class. Even still, I think the benefits are doubtful.

Reply:
I think there's clearly value in having the class, private or not, for code reuse, usability and testability (the latter two coming from having a common interface). I'm not strongly opposed to making it private, but can think of two reasons not to. First, for type checking, though this could be achieved by defining (publicly) something like `FERMIONIC_SIMULATION_GATES = (QuadraticFermionicSimulationGate, ...)`. Second, while at this point the main goal is to refactor the three existing fermionic simulation gates, there may be other use cases. For example, one might want to define an `ExtendedQuadraticFermionicSimulationGate` that simulates two modes that are _not_ adjacent in the JW string; this would allow, e.g., specialized unitary and decomposition code.

Reply:
If we make FermionicSimulationGate a public class, we should decide, and specify, what a FermionicSimulationGate actually is.

Reply:
What is insufficient about the current docstring: "The Jordan-Wigner transform of exp(-i H) for a fermionic Hamiltonian H."

Reply:
That doesn't specify any restriction on H, which means H could be any fermionic Hamiltonian, which means the gate could be any fermionic unitary. But I've described above some potential problems with that (the actual matrix is undefined without specifying the ordering of fermions), and furthermore you already stated above that that is not the intent:
> The idea is definitely not to represent the JW transform of any fermionic unitary. The goal here is specifically to refactor the three existing gates (`QuadraticFermionicSimulationGate`, `CubicFermionicSimulationGate`, `QuarticFermionicSimulationGate`), so that they have a common interface and share code to the extent possible.